lib/codegen/DecisionParser.cpp

```diff
@@ -0,0 +1,154 @@
+      constructors.push_back(binding);
+
+      parents.push_back("");
+      auto child = (*this)(node["next"]); 
+      parents.pop_back();
+
+      auto result = FunctionNode::Create(binding, function, child, cat);
+      
+      YAML::Node vars = node["args"];
+      for (auto iter = vars.begin(); iter != vars.end(); ++iter) {
+        auto var = *iter;
+        int idx1 = var[0].as<int>();
+        int idx2 = var[1].as<int>();
+        if (idx1 == 0) {
+          result->addBinding(constructors[constructors.size()-2-idx2]);
+        } else {
+          result->addBinding(parents[parents.size()-idx1]);
+        }
+      }
+      return result;
+    }
+    case SwitchLit:
+    case Switch: {
+      YAML::Node list = node["specializations"];
+      std::string name = constructors.back();
+      constructors.pop_back();
+      auto result = SwitchNode::Create(name);
+      parents.push_back(name);
+      for (auto iter = list.begin(); iter != list.end(); ++iter) {
+        auto _case = *iter;
+        std::vector<std::string> copy = constructors;
+        std::vector<std::string> bindings;
+        KOREObjectSymbol *symbol;
+        if (kind == SwitchLit) {
+          symbol = dv;
+        } else {
+          std::string symName = _case[0].as<std::string>();
+          symbol = syms.lookup(symName);
+          for (int i = 0; i < symbol->getArguments().size(); ++i) {
+            std::string binding = "_" + std::to_string(counter++);
+            constructors.push_back(binding);
+            bindings.push_back(binding);
+          }
+          std::reverse(constructors.end()-symbol->getArguments().size(), constructors.end());
+        }
+        DecisionNode *child = (*this)(_case[1]);
+        constructors = copy;
+        if (kind == SwitchLit) {
+          int bitwidth = node["bitwidth"].as<int>();
+          result->addCase({symbol, {bitwidth, _case[0].as<std::string>(), 10}, child}); 
+        } else {
+          result->addCase({symbol, bindings, child});
+        }
+      }
+      auto _case = node["default"];
+      if (!_case.IsNull()) {
+        std::vector<std::string> copy = constructors;
+        DecisionNode *child = (*this)(_case);
+        constructors = copy;
+        result->addCase({nullptr, std::vector<std::string>{}, child});
+      }
+      parents.pop_back();
+      return result;
+    }
+    case Leaf: {
+      int action = node["action"][0].as<int>();
+      std::string name = "apply_rule_" + std::to_string(action);
+      auto result = LeafNode::Create(name);
+      YAML::Node vars = node["action"][1];
+      for (auto iter = vars.begin(); iter != vars.end(); ++iter) {
+        auto var = *iter;
+        int idx1 = var[0].as<int>();
+        int idx2 = var[1].as<int>();
+        if (idx1 == 0) {
+          result->addBinding(constructors[constructors.size()-1-idx2]);
+        } else {
+          result->addBinding(parents[parents.size()-idx1]);
+        }
+      }
+      return result;
+    }
+    }     
+  }
+};
```

Comment:
I kind of feel that each of these cases should be put in separate functions due to the length of this, but it's a nit.